src/charm.py

```diff
@@ -0,0 +1,256 @@
+
+from jinja2 import Environment, FileSystemLoader
+from datetime import timedelta
+from time import sleep
+
+
+class CockroachStartedEvent(EventBase):
+    pass
+
+
+class ClusterInitializedEvent(EventBase):
+    def __init__(self, handle, cluster_id):
+        super().__init__(handle)
+        self.cluster_id = cluster_id
+
+    def snapshot(self):
+        return self.cluster_id
+
+    def restore(self, cluster_id):
+        self.cluster_id = cluster_id
+
+
+class CockroachDBCharmEvents(CharmEvents):
+    cockroachdb_started = EventSource(CockroachStartedEvent)
+    cluster_initialized = EventSource(ClusterInitializedEvent)
+
+
+class CockroachDBCharm(CharmBase):
+    on = CockroachDBCharmEvents()
+    state = StoredState()
+
+    COCKROACHDB_SERVICE = 'cockroachdb.service'
+    SYSTEMD_SERVICE_FILE = f'/etc/systemd/system/{COCKROACHDB_SERVICE}'
+    WORKING_DIRECTORY = '/var/lib/cockroach'
+    COCKROACH_INSTALL_DIR = '/usr/local/bin'
+    COCKROACH_BINARY_PATH = f'{COCKROACH_INSTALL_DIR}/cockroach'
+    COCKROACH_USERNAME = 'cockroach'
+    PSQL_PORT = 26257
+    HTTP_PORT = 8080
+
+    MAX_RETRIES = 10
+    RETRY_TIMEOUT = timedelta(milliseconds=125)
+
+    def __init__(self, framework, key):
+        super().__init__(framework, key)
+
+        self.state.set_default(is_started=False)
+
+        for event in (self.on.install,
+                      self.on.start,
+                      # self.on.upgrade_charm,
+                      self.on.config_changed,
+                      self.on.cluster_relation_changed,
+                      self.on.cockroachdb_started,
+                      self.on.proxy_listen_tcp_relation_joined):
+            self.framework.observe(event, self)
+
+        self.cluster = CockroachDBCluster(self, 'cluster')
+        self.tcp_load_balancer = ProxyListenTcpInterfaceProvides(self, 'proxy-listen-tcp')
+
+    def on_install(self, event):
+        try:
+            resource_path = self.model.resources.fetch('cockroach-linux-amd64')
+        except ModelError:
+            resource_path = None
+
+        if resource_path is None:
+            ARCHITECTURE = 'amd64'  # hard-coded until it becomes important
+            version = self.model.config['version']
+            cmd = (f'wget -qO- https://binaries.cockroachdb.com/'
+                   f'cockroach-{version}.linux-{ARCHITECTURE}.tgz'
+                   f'| tar -C {self.COCKROACH_INSTALL_DIR} -xvz --wildcards'
+                   ' --strip-components 1 --no-anchored "cockroach*/cockroach"')
+            subprocess.check_call(cmd, shell=True)
+            os.chown(self.COCKROACH_BINARY_PATH, 0, 0)
+        else:
+            cmd = ['tar', '-C', self.COCKROACH_INSTALL_DIR, '-xv', '--wildcards',
+                   '--strip-components', '1', '--no-anchored', 'cockroach*/cockroach',
+                   '-zf', str(resource_path)]
+            subprocess.check_call(cmd)
+
+        self._setup_systemd_service()
+
+    @property
+    def is_single_node(self):
```

Comment:
Stopped that initial review here.

Reply:
New changes:

https://github.com/canonical/cockroachdb-operator/pull/1/commits/e6741a9a01432f19d445914a77ff3a65050f0bcd